### PR TITLE
Enable rocm for tensorflow. Add needed ROCm components

### PR DESCRIPTION
**Pull request by srekolam**
This PR adds the rocm components needed to build tensorflow for rocm.
This was tested using  rocm-5.1.0 binaries at /opt/rocm/rocm-5.1.0. The packages.yaml file was basically updated to point to these binaries.
A symbolic link inside /opt called rocm that points to rocm-5.1.0 is required.
(ln -s /opt/rocm-5.1.0 rocm)

After the build was successful , I ran the testcase which gave the output as attached on a machine with 8 MI50's.

The addition of the below was required for the protobuf error and was provided by @lukebroskop 
filter_file(
            '"-U_FORTIFY_SOURCE",',
            "\"-U_FORTIFY_SOURCE\", \"-I%s\"," % self.spec['protobuf'].prefix.include,
            "third_party/gpus/crosstool/BUILD.rocm.tpl")

[tensorflow_testcase.txt](https://github.com/spack/spack/files/9015709/tensorflow_testcase.txt)
[packages.yaml.txt](https://github.com/spack/spack/files/9015711/packages.yaml.txt)

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-tensorflow/package.py [hunk 1]
```diff
@@ -617,6 +633,10 @@ def setup_build_environment(self, env):
             env.set('INCLUDEDIR', spec['protobuf'].prefix.include)
 
     def patch(self):
+        filter_file(
+            '"-U_FORTIFY_SOURCE",',
+            "\"-U_FORTIFY_SOURCE\", \"-I%s\"," % self.spec['protobuf'].prefix.include,
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-tensorflow/package.py]:
```suggestion
            '"-U_FORTIFY_SOURCE", "-I%s",' % self.spec['protobuf'].prefix.include,
```

**Comment by adamjstewart** [comment]
Has this patch been contributed upstream?

**Comment by adamjstewart** [comment]
How is this PR related to #23431?

File: var/spack/repos/builtin/packages/py-tensorflow/package.py (hunk 1 above)
Comment (srekolam) [reply var/spack/repos/builtin/packages/py-tensorflow/package.py]: i have taken the changes suggested by you.

**Comment by srekolam** [comment]
> How is this PR related to #23431?

the original changes that were made as part of #23431 were intended to work with spack installed rocm binaries. But i ran into build issues as build were looking for headers from the /opt/rocm path and so, i was seeing some issues. The branch in #23431 is a working branch and was recently updated with patch changes provided internal tensorflow team. This would again depend on the binaries inside /opt/rocm due to current implementation inside third_party/gpus/rocm_configure.bzl etc. I intend to work with the team so that these constraints get addressed in the future releases .

**Comment by srekolam** [comment]
> Has this patch been contributed upstream?

Not yet . The patch represents the differences between rocm 's tensorflow fork and upstream -2.7.0 version.  I think (needs reconfirmation from the internal tensorflow team on the timeline) these changes may eventually become part of the upstream .

**Comment by srekolam** [comment]
@adamjstewart , let me know if you have any review comments .

**Comment by adamjstewart** [comment]
I mostly still have the same questions. 

> How is this PR related to #23431?

It seems like this PR is almost a duplicate of #23431. Should we close one of these PRs and focus on the other?

> Has this patch been contributed upstream?

This is a really big patch, I don't know anything about bazel, and I don't have access to an AMD GPU to test this on. For me, all I can say is "this is indeed a patch that does something", which isn't a very helpful review. For my sake and also for your own sake, I would love to get a review of this patch from an expert. If you submit this PR upstream, I trust the TF devs much more than I trust myself to review this.

I've seen a lot of patches contributed to Spack that never get contributed upstream and someone has to keep updating them with every new release. I don't want to maintain something like that. I would rather solve the problem in TF first and then backport merged patches if needed.

From my side, in order for this PR to move forward, I would like to first close #23431 and wait for this patch to be approved and merged by a TF developer.

**Comment by srekolam** [comment]
> I mostly still have the same questions.
> 
> > How is this PR related to #23431?
> 
> It seems like this PR is almost a duplicate of #23431. Should we close one of these PRs and focus on the other?
> 
> > Has this patch been contributed upstream?
> 
> This is a really big patch, I don't know anything about bazel, and I don't have access to an AMD GPU to test this on. For me, all I can say is "this is indeed a patch that does something", which isn't a very helpful review. For my sake and also for your own sake, I would love to get a review of this patch from an expert. If you submit this PR upstream, I trust the TF devs much more than I trust myself to review this.
> 
> I've seen a lot of patches contributed to Spack that never get contributed upstream and someone has to keep updating them with every new release. I don't want to maintain something like that. I would rather solve the problem in TF first and then backport merged patches if needed.
> 
> From my side, in order for this PR to move forward, I would like to first close #23431 and wait for this patch to be approved and merged by a TF developer.

Thanks @adamjstewart  .i have closed #23431 . I will check with my internal TF team with regard to the plan of submitting the patch changes for the upstream.

**Comment by srekolam** [comment]
@adamjstewart , this is what i have received today - The TF  team from AMD wants to enable the tip of the upstream TF repo to build with ROCm (this would eliminate the need for this patch file) and/but details are further awaited .. Also , If you can provide inputs on the need to support TF versions 2.9.0, 2.8.0 and 2.7.0 , TF team internally can look at options in future .

**Comment by adamjstewart** [comment]
We definitely want to add TF 2.8 and 2.9 to the package. If the patch is accepted upstream, I'm happy to add backports of the patch to TF 2.7, 2.8, and 2.9. 

P.S. We're currently experiencing build issues for TF 2.7 (https://github.com/spack/spack/issues/31229) so if your team has any idea how to resolve those let me know.

**Comment by srekolam** [comment]
@adamjstewart i am closing the PR and submitting a new PR. Basically as we discussed the concerns  with the Patch ([tf_27_rocm_changes_to_support_without_creation_of_softlink.patch](https://github.com/spack/spack/pull/31380/files#diff-e4c1d93d785dc9ca8ff76fac92bf007b12b9de0e0823445a082098f98c99d70f)  and how this cannot be soln for future releases, I made some changes to use the rocm fork instead for older tf release ( 2.7.4)  and see if that can help us move forward as a interim solution till i can use the upstream and build rocm with out any patches etc. My new PR will hopefully be acceptable .

**Pull request closed without merging**